query/src/main/java/com/connexta/replication/query/QueryManager.java

```diff
@@ -0,0 +1,128 @@
+
+  /** Startup delay for polling for configuration changes in second. */
+  private static final long STARTUP_DELAY = TimeUnit.MINUTES.toSeconds(1);
+
+  private final QueryServiceTools queryServiceTools;
+
+  /** Sets of identifiers for the sites to be managed or empty to manage all sites. */
+  private final Set<String> sites;
+
+  private final ScheduledExecutorService executor;
+
+  /** Sets of active query services keyed by site its. */
+  @VisibleForTesting final Map<String, QueryService> services = new ConcurrentHashMap<>();
+
+  public QueryManager(Stream<String> sites, QueryServiceTools queryServiceTools) {
+    this(sites, queryServiceTools, Executors.newSingleThreadScheduledExecutor());
+  }
+
+  /**
+   * Instantiate a new query manager.
+   *
+   * @param sites identifiers for the sites to be managed or empty to manage all sites
+   */
+  QueryManager(
+      Stream<String> sites,
+      QueryServiceTools queryServiceTools,
+      ScheduledExecutorService executor) {
+    this.sites = sites.collect(Collectors.toSet());
+    this.queryServiceTools = queryServiceTools;
+    this.executor = executor;
+  }
+
+  /**
+   * Initializes the query manager and kick start all query services to start polling sites that
+   * must be polled.
+   */
+  public void init() {
+    if (sites.isEmpty()) {
+      LOGGER.info("Managing queries for all configured sites.");
+    } else {
+      LOGGER.info("Managing queries for sites: {}.", sites);
+    }
+    executor.scheduleAtFixedRate(
+        this::reloadSiteConfigs,
+        QueryManager.STARTUP_DELAY,
+        queryServiceTools.getPeriod(),
+        TimeUnit.SECONDS);
+  }
+
+  /** Destroys the query manager and stop all currently running query services. */
+  public void destroy() {
+    LOGGER.info("Shutting down query managers");
+    executor.shutdownNow();
+    for (QueryService service : services.values()) {
+      service.stop();
+    }
+  }
+
+  void reloadSiteConfigs() {
+    final Map<String, Site> newSites =
+        queryServiceTools.allSites().collect(Collectors.toMap(Site::getId, Function.identity()));
+
+    // stop query services for sites that are no longer configured or have changed to a non-polled
+    // type
+    // update them if they are already running
+    updateQueryServices(newSites);
+
+    // start query services for new sites that must be polled
+    newSites.values().stream()
+        .filter(QueryManager::mustBePolled)
+        .map(this::newService)
+        .forEach(QueryService::start);
+  }
+
+  void updateQueryServices(Map<String, Site> newSites) {
+    for (Entry<String, QueryService> entry : services.entrySet()) {
+      final Site updatedSite = newSites.remove(entry.getKey());
+
+      if (updatedSite != null && updatedSite.getType().mustBePolled()) {
+        entry.getValue().update(updatedSite);
+      } else {
+        services.remove(entry.getKey());
+        entry.getValue().stop();
```

Comment:
❗️ All sites that are stopped should also be removed from `services`

Reply:
That's what's happening in the line above but this comment tells me the code isn't very readable. Perhaps just switching the order of the lines would help?

Reply:
I did miss that. I would be careful about removing from a collection you are iterating without going through an actual iterator. This can tend to lead to concurrent modification exceptions. The best approach is to get an iterator on the entry set and iterate using that. When you want to remove the entry, you simply call `i.remove()`. This is more optimal since the iterator is already tracking where it is and avoid the concurrent modification exception that can occur.